10_rest/runnerdave/pkg/puppy/rest_test.go

```diff
@@ -0,0 +1,224 @@
+package puppy_test
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	puppy "github.com/anz-bank/go-course/10_rest/runnerdave/pkg/puppy"
+	store "github.com/anz-bank/go-course/10_rest/runnerdave/pkg/puppy/store"
+)
+
+var (
+	rs puppy.RestServer
+)
+
+func TestGetPuppy(t *testing.T) {
+	req, err := http.NewRequest("GET", "/api/puppy/1", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	rr := httptest.NewRecorder()
+	newPuppy := puppy.Puppy{
+		Breed:  "ULTRASURE",
+		Colour: "green",
+		Value:  2732.81,
+	}
+	db := store.NewSyncStore()
+	if err := db.CreatePuppy(newPuppy); err != nil {
+		panic(err)
+	}
+	handler := rs.SetupRoutes(db)
+
+	handler.ServeHTTP(rr, req)
+
+	if status := rr.Code; status != http.StatusOK {
+		t.Errorf("handler returned wrong status code: got %v want %v",
+			status, http.StatusOK)
+	}
+
+	expected := `{
+		"id": 1,
+		"breed": "ULTRASURE",
+		"color": "green",
+		"value": 2732.81
+	  }`
+	actual := rr.Body.String()
+	require.JSONEq(t, expected, actual, "GET Puppy was incorrect, got: %s, want: %s.",
+		actual, expected)
+}
+
+func TestGetPuppyNotFound(t *testing.T) {
+	req, err := http.NewRequest("GET", "/api/puppy/200", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	rr := httptest.NewRecorder()
+	handler := rs.SetupRoutes(store.NewSyncStore())
+
+	handler.ServeHTTP(rr, req)
+
+	if status := rr.Code; status != http.StatusNotFound {
+		t.Errorf("handler returned wrong status code: got %v want %v",
+			status, http.StatusOK)
+	}
+
+	expected := `{
+		"id": 0,
+		"breed": "",
+		"value": 0
+	  }`
+	actual := rr.Body.String()
+	require.JSONEq(t, expected, actual, "GET Puppy was incorrect, got: %s, want: %s.",
+		actual, expected)
+}
+
+func TestPostPuppy(t *testing.T) {
+	newPuppy := puppy.Puppy{
+		ID:     1,
+		Breed:  "EPLODE",
+		Colour: "brown",
+		Value:  3889.92,
+	}
+	body, _ := json.Marshal(newPuppy)
```

Comment:
Don't ignore errors, especially in tests - it leads to false confidence that all is OK. Since this is setting up the test, you should use `require.NoError(err)`. Same in the other places you are ignoring errors. Usually you would never expect these errors to fire, but every now and then, something changes you did not expect, and it is very useful to have them then.

Reply:
fixed